porch/pkg/meta/store.go

```diff
@@ -0,0 +1,229 @@
+// stores metadata in a CRD.
+type crdMetadataStore struct {
+	coreClient client.Client
+}
+
+func (c *crdMetadataStore) Get(ctx context.Context, namespacedName types.NamespacedName) (PackageRevisionMeta, error) {
+	ctx, span := tracer.Start(ctx, "crdMetadataStore::Get", trace.WithAttributes())
+	defer span.End()
+
+	var internalPkgRev internalapi.PackageRev
+	err := c.coreClient.Get(ctx, namespacedName, &internalPkgRev)
+	if err != nil {
+		return PackageRevisionMeta{}, err
+	}
+
+	labels := internalPkgRev.Labels
+	delete(labels, PkgRevisionRepoLabel)
+
+	return PackageRevisionMeta{
+		Name:        internalPkgRev.Name,
+		Namespace:   internalPkgRev.Namespace,
+		Labels:      labels,
+		Annotations: internalPkgRev.Annotations,
+	}, nil
+}
+
+func (c *crdMetadataStore) List(ctx context.Context, repo *configapi.Repository) ([]PackageRevisionMeta, error) {
+	ctx, span := tracer.Start(ctx, "crdMetadataStore::List", trace.WithAttributes())
+	defer span.End()
+
+	var internalPkgRevList internalapi.PackageRevList
+	err := c.coreClient.List(ctx, &internalPkgRevList, client.InNamespace(repo.Namespace), client.MatchingLabels(map[string]string{PkgRevisionRepoLabel: repo.Name}))
+	if err != nil {
+		return nil, err
+	}
+	var pkgRevMetas []PackageRevisionMeta
+	var names []string
+	for _, ipr := range internalPkgRevList.Items {
+		labels := ipr.Labels
+		delete(labels, PkgRevisionRepoLabel)
+		pkgRevMetas = append(pkgRevMetas, PackageRevisionMeta{
+			Name:        ipr.Name,
+			Namespace:   ipr.Namespace,
+			Labels:      labels,
+			Annotations: ipr.Annotations,
+		})
+		names = append(names, ipr.Name)
+	}
+	return pkgRevMetas, nil
+}
+
+func (c *crdMetadataStore) Create(ctx context.Context, pkgRevMeta PackageRevisionMeta, repo *configapi.Repository) (PackageRevisionMeta, error) {
+	ctx, span := tracer.Start(ctx, "crdMetadataStore::Create", trace.WithAttributes())
+	defer span.End()
+
+	labels := pkgRevMeta.Labels
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[PkgRevisionRepoLabel] = repo.Name
+	internalPkgRev := internalapi.PackageRev{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        pkgRevMeta.Name,
+			Namespace:   pkgRevMeta.Namespace,
+			Labels:      labels,
+			Annotations: pkgRevMeta.Annotations,
+			// We probably should make these owner refs point to the PackageRevision CRs instead.
+			// But we need to make sure that deletion of these are correctly picked up by the
+			// GC. Currently we delete PackageRevisions through polling of the git/oci repos, and
+			// that doesn't get picked up by the GC.
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: configapi.RepositoryGVK.GroupVersion().String(),
+					Kind:       configapi.RepositoryGVK.Kind,
+					Name:       repo.Name,
+					UID:        repo.UID,
+				},
+			},
+		},
+	}
+	if err := c.coreClient.Create(ctx, &internalPkgRev); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return c.Update(ctx, pkgRevMeta)
```

Comment:
Nit: ideally we would not do this, as it indicates that we probably aren't doing optimistic concurrency correctly.  But, fine to tackle later

Reply:
Yeah, the optimistic concurrency is a bit tricky with the two separate sources.